Add recipe for code-cells

astoff opened the pull request:
### Brief summary of what the package does

This package does two related but technically independent things. First, it provides utilities to edit code split into cells, like the python-cell package already does, but in a language-independent way.

The second function is to allow editing Jupyter (ipynb) notebooks. These are JSON files, and the package installs a major-mode-alist entry to convert them to a regular script (using an external tool) when the file is opened, and back to ipynb format when saving.

### Direct link to the package repository

https://github.com/astoff/cells.el/

### Your association with the package

I'm the maintainer.

### Relevant communications with the upstream package maintainer

None needed

### Checklist

Please confirm with `x`:

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [ ] My elisp byte-compiles cleanly
    - I get the warning `cells.el:107:1:Warning: Unused lexical variable ‘using-region’`, but this is exactly what I want to do. That variable is supposed to be available to the user inside a macro body.

- [X] `M-x checkdoc` is happy with my docstrings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

riscy commented:
Just a heads up, the CI failure in this case is innocuous and can be disregarded.

alphapapa commented:
FYI, there is an Emacs library named `cell`.  Not that it would necessarily justify renaming your package, but something to consider.  http://xelf.me/cell.html

astoff commented:
I'm not sure the name similarity warrants renaming my package, but if you insist a bit more I'm open to renaming it to `code-cells`. Perhaps this is actually a more descriptive name.

By the way, since the package hooks into the buffer saving process, which is a quite delicate thing, I would appreciate any feedback. Specifically, should the `cells-write-ipynb` function take any additional precautions to be used in `write-file-functions`?

alphapapa commented:
> I'm not sure the name similarity warrants renaming my package, but if you insist a bit more I'm open to renaming it to `code-cells`. Perhaps this is actually a more descriptive name.

I do think that name is more descriptive, so I'd recommend it, but I have no authority to insist.  :)

> By the way, since the package hooks into the buffer saving process, which is a quite delicate thing, I would appreciate any feedback. Specifically, should the `cells-write-ipynb` function take any additional precautions to be used in `write-file-functions`?

If you're asking me, I don't know, as I haven't worked with `write-file-functions` before.  The only thing I could say would be that, generally, the code should be careful to handle any errors that might arise, because errors while saving might interfere with Emacs's normal operation--but I'm sure you already know that.  :)

riscy commented:
This looks cool.  I liked the name `code-cells` but I think sticking with `cells` is fine.  Just make sure your summary line has disambiguating words in it, which it looks like it already does.

Considering that the code does two separate things, have you thought about splitting this into two packages, possibly with one dependent on the other?  I can imagine someone would enjoy the ipynb editing facilities, but might prefer to use `python-cell` for their notebooking.  This isn't a blocker, but if you can see these facilities growing in different directions it might be worth it to do this early.

Here's the latest checklist:

### cells.el ###

byte-compile (using Emacs 27.1):
```
cells.el:64:1:Error: Symbol's value as variable is void: cells-boundary-markers
```

checkdoc (using version 0.6.2):
```
Warning (emacs): 
cells.el:116: Argument `docstring' should appear (as DOCSTRING) in the doc string
Warning (emacs): 
cells.el:157: Disambiguate outline-level by preceding w/ function,command,variable,option or symbol.
```

package-lint (using version 20200906.512):
```
1 issue found:
1:0: warning: "Version:" or "Package-Version:" header is missing. MELPA will handle this, but other archives will not.
```

Other [lints](https://github.com/riscy/melpazoid):
- cells.el#L233: It's safer to sharp-quote function names; use `#'`
- cells.el#L21: Prefer `https` over `http` if possible ([why?](https://news.ycombinator.com/item?id=22933774))

astoff commented:
All right, I just renamed the package.

I had also thought about splitting it up in two, but then instead of one small package, we would have two tiny packages. Too much fragmentation also seems something to avoid, and since `code-cells` would be the dependent package anyway, I think the splitting can be done later if the occasion arises.

I don't get the error at cells.el:64:1 when byte-compiling locally, and I don't get what that means -- cells-boundary-markers is not a function, and doesn't anywhere it could be mistaken for such.

riscy commented:
All good, keeping it in one package is fine.

I'm not entirely sure what happened, but I think you merged master onto this branch before pushing the name change, which created duplicate commits and a large diff for this PR.  Would you be willing to clean that up?  E.g. rebase, squash, whatever suits you.  Otherwise I'd have to check very carefully before merging.

The reason you may not be seeing the byte-compilation error is probably because you evaluated the buffer in your Emacs session before compilation.  `rx` is a macro which the byte compiler attempts to macroexpand before `code-cells-boundary-markers` has a chance to be declared.  This also reveals a small bug which is that, even if the value of `code-cells-boundary-markers` is changed to a new value, `code-cells-boundary-regexp` will still return results reflecting the old value.

astoff commented:
Ok, the git clumsiness is fixed. Thanks for the heads-up about the macro thing, will fix it as well.

riscy commented:
Feel free to ping when ready for another look!

astoff commented:
From my perspective, this is ready to merge. I have also fixed the `rx` compilation issues you pointed out above.

riscy commented:
Here's the latest --

### code-cells.el ###

byte-compile (using Emacs 27.1):
```
code-cells.el:108:1:Warning: Unused lexical variable `using-region'
In code-cells-mode:
code-cells.el:196:50:Warning: reference to free variable
    `outline-heading-end-regexp'
code-cells.el:204:17:Warning: assignment to free variable
    `outline-heading-end-regexp'
```

checkdoc (using version 0.6.2):
```
Warning (emacs): 
code-cells.el:117: Argument `docstring' should appear (as DOCSTRING) in the doc string
Warning (emacs): 
code-cells.el:160: Disambiguate outline-level by preceding w/ function,command,variable,option or symbol.
```

package-lint (using version 20200906.512):
```
1 issue found:
1:0: warning: "Version:" or "Package-Version:" header is missing. MELPA will handle this, but other archives will not.
```
Prefer `Version`

astoff commented:
Ok, I've addressed most of the issues. I believe the two remaining ones are false positives:

```
code-cells.el:108:1:Warning: Unused lexical variable `using-region'
```
This is variable is not used within the library, but it' supposed to be available for the user in the body of the `code-cells-do` macro. Is there anything I can do to silence this warning?

```
code-cells.el:160: Disambiguate outline-level by preceding w/ function,command,variable,option or symbol.
```
Here I think the docstring is clear, and changing it to the expected format would be awkward.

riscy commented:
> This is variable is not used within the library, but it' supposed to be available for the user in the body of the code-cells-do macro. Is there anything I can do to silence this warning?

I believe if you forward-declare it with `(defvar using-region)` or ideally `(defvar code-cells-using-region)` -- unless `using-region` is some idiom I'm not aware of -- then that will let the byte-compiler know you intend this.

> Here I think the docstring is clear

This checkdoc can be annoying, but by putting that symbol in matched quotes you're telling Emacs that you want it to create a link in the online help -- and it doesn't know where to send the user since there is both a builtin variable variable and a builtin function with that name.  Personally I would make that sentence the last line of the docstring and just write something like "Note this is the `code-cells-mode' analog of the function `outline-level'."  This isn't strictly a blocker but if you're 100% green on checkdoc then you'll notice more easily when something important goes wrong.

astoff commented:
All right, I've addressed these two last issues. Thanks for all the suggestions!

astoff commented:
@riscy Friendly ping :-)

riscy commented:
Should be back to take a look on the weekend at the latest, if not sooner!

riscy commented:
Looks good, thanks for this!

This will actually be at least the second package on MELPA that adds a major mode for editing .ipynb files.  It might be worth adding a note to your users, in the README or Commentary, that your package could conflict if with [ein](https://melpa.org/#/ein) if that package is installed as well (both add to `auto-mode-alist`).  I'll mention @dickmao here as well, who seems to be doing a lot of ein maintenance recently.
